install/kubernetes: remove privileged from Cilium DaemonSet

aanm opened the pull request:
We are already setting all the capabilities that Cilium requires so we
don't need to set privileged as true.

Signed-off-by: André Martins <andre@cilium.io>

```release-note
Remove privileged mode in Cilium's DaemonSet
```

**Note for reviewers**: all tests have passed on PR https://github.com/cilium/cilium/pull/19449 which includes these changes.

aanm committed:
install/kubernetes: remove privileged from Cilium DaemonSet

Cilium container does not require to run as privileged so we can
effectively remove "privileged: true" and set a minimal set of
capabilities required for Cilium to run.

Signed-off-by: André Martins <andre@cilium.io>

sayboras commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -372,7 +372,6 @@ spec:
           capabilities:
             add:
             - NET_ADMIN
-          privileged: true
```
maybe we can have a quick look for this secion https://github.com/cilium/cilium/blame/master/Documentation/operations/system_requirements.rst#L387-L396 and check if any change is required.

aanm replied:
Good spotting, yes we are looking into the capabilities that Cilium needs. @borkmann the docs says `CAP_SYS_ADMIN` is needed but Cilium only have `NET_ADMIN`, does `NET_ADMIN` also have `SYS_ADMIN`?

sayboras reviewed (approved):
LGTM 💯 , one minor comment on docs.

kaworu commented on bpf/init.sh:
```diff
@@ -14,30 +14,32 @@
+LIB=${1}
+RUNDIR=${2}
+PROCSYSNETDIR=${3}
+SYSDIR=${4}
```
nit: Since we're only using `${SYSDIR}/class/net`, the "asymmetry" with `${PROCSYSNETDIR}` looks odd to me. In other words, either one of:

* `${PROCDIR}` for `/proc` and `${SYSDIR}` for `/sys`
* `${PROCSYSNETDIR}` for `/proc/sys/net` and `${SYSCLASSNETDIR}` for `/sys/class/net`

would "feel cleaner", does that make sense?

errordeveloper commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -230,13 +231,51 @@ spec:
+              - SYS_RESOURCE
+              # Re-add if endpoint create fails or if the log driver is using syslog
+              #- SYSLOG
+              # Requires kernel 5.8 and container runtime X
```
What's X?

errordeveloper commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -230,13 +231,51 @@ spec:
+          name: host-proc-sys-kernel
 {{- /* CRI-O already mounts the BPF filesystem */ -}}
 {{- if not (eq .Values.containerRuntime.integration "crio") }}
-        - mountPath: /sys/fs/bpf
```
This looks like a cosmetic change, but it certainly deserves a separate commit!

errordeveloper commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -230,13 +231,51 @@ spec:
+              #- PERFMON
+              #- BPF
         volumeMounts:
+        - mountPath: /host/proc/sys/net
```
So with `privileged: true` the pod was getting `/proc/sys/net` and `/proc/sys/kernel` automatically? I'd say that it needs to be stated more explicitly in the commit message, as it looks like a significant change and not at all obvious. I would even consider putting it in a separate commit also, which can go before the change to `securityContext`.

errordeveloper reviewed (changes requested):
This is really cool, and I think this is complex enough to be broken into 3 commits like this:

1) bind-mount `/proc/sys` dirs explicitly from the host (prepare to remove privileged mode)
2) use least privileges for agent container
3) change path to BPF filesystem for consistency

JAORMX commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -233,13 +239,53 @@ spec:
         securityContext:
+          {{- if .Values.securityContext.privileged }}
           privileged: true
+          {{- else }}
```
if the `privileged` flag is not set, you can disable make the container "unconfined" for SELinux (which would be needed here), by setting the following:

```
securityContext:
  seLinuxOptions:
    level: 's0'
    type: 'spc_t'
```

The level might not have been strictly needed here, but it's good to add just in case.

aanm committed:
install/kubernetes: remove privilege mode from node-init

Similarly to what was done for the Cilium DaemonSet, the node-init
DaemonSet does not require so many permissions and we can trimmed them
down.

Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -825,6 +825,7 @@ data:
   enable-bgp-control-plane: "false"
 {{- end }}
 
+  procfs: {{ .Values.procfs | quote }}
```
I would have expected that the user may want to choose where the _host_ procfs is, so that we can map that into the container. I don't understand why we would let the user configure the path _inside_ the container, since there's only one correct configuration parameter for this value (otherwise it's incompatible with the hardcoded changes in the `DaemonSet`).

aanm replied:
Oh, very good catch, it does not make sense to expose this option to the user. It should be this logic instead:
```
{{- if not .Values.securityContext.privileged }}
  procfs: "/host/proc"
{{- end }}
```

joestringer commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -233,13 +241,59 @@ spec:
+          {{- if .Values.securityContext.privileged }}
           privileged: true
+          {{- else }}
+          seLinuxOptions:
+            level: 's0'
+            # Running with spc_t since we have removed the privileged mode.
+            # Users can change it to a different type as long as they have the
+            # type available on the system.
+            type: 'spc_t'
```
Does this mean that privileged=false requires SELinux?

aanm replied:
No no, it only means that if SELinux is available / enabled then we should be running the container with level `s0` and type `spc_t`

joestringer commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -362,7 +416,45 @@ spec:
+        - --
+        securityContext:
+          privileged: true
+      {{- /* CRI-O already mounts the BPF filesystem */ -}}
+      {{- if not (eq .Values.containerRuntime.integration "crio") }}
```
This part seems like it should be above the whole init container that's defined here..?

aanm replied:
Good point, however I don't think it hurts running this init container even if cri-o already mounts the BPF filesystem. Let's assume worst case where cri-o mounts the directory inside the container but the bpf fs is not mounted.

kaworu commented on pkg/sysctl/sysctl.go:
```diff
@@ -54,7 +60,31 @@ func parameterPath(name string) (string, error) {
+func SetProcfs(path string) {
+	if path == procFsDefault {
+		return
+	}
```
Can you elaborate the rational behind this codepath? I find it confusing that `SetProcfs(procFsDefault)` is consistently a no-op:
```golang
SetProcfs("/foo")
SetProcfs(procFsDefault)
p := GetProcfs() // here p == "/foo"
SetProcfs(procFsDefault) // no panic
SetProcfs("/foo") // panic although could be idem-potent
```

aanm replied:
Yeah I don't see any rational behind it, this code was written over 1y ago. Removed.

kaworu reviewed (commented):
One small question. I'm deferring review here as I'm lacking context.

joestringer commented on Documentation/helm-values.rst:
```diff
@@ -1381,6 +1381,10 @@
+   * - procfs
+     - 
+     - string
+     - ``"/host/proc"``
```
Has this file been regenerated? I think you removed this option, and it shouldn't need to be exposed to the user any more...?

joestringer reviewed (approved):
Just one nit, seems like the helm-values docs page didn't get updated after the last change. Other than that, LGTM

kaworu commented on pkg/sysctl/sysctl.go:
```diff
@@ -54,7 +60,28 @@ func parameterPath(name string) (string, error) {
+}
+
+// SetProcfs sets path for the root's /proc. Setting this value with a different
+// default value and executing it after GetProcfs causes panic.
```
This comment is still referring to the previous behavior
```suggestion
// SetProcfs sets path for the root's /proc. Calling it after GetProcfs causes panic.
```
